topovis/topologytoolkit/src/processors/morsesmalecomplex.cpp

```diff
@@ -0,0 +1,147 @@
+
+// The Class Identifier has to be globally unique. Use a reverse DNS naming scheme
+const ProcessorInfo MorseSmaleComplex::processorInfo_{
+    "org.inviwo.MorseSmaleComplex",      // Class identifier
+    "Morse-Smale Complex",               // Display name
+    "Topology",                          // Category
+    CodeState::Experimental,             // Code state
+    "CPU, Topology, TTK, Contour Tree",  // Tags
+};
+const ProcessorInfo MorseSmaleComplex::getProcessorInfo() const { return processorInfo_; }
+
+MorseSmaleComplex::MorseSmaleComplex()
+    : Processor()
+    , inport_{"triangulation"}
+    , outport_{"outport"} {
+    addPort(inport_);
+    addPort(outport_);
+}
+
+void MorseSmaleComplex::process() {
+    if (util::is_future_ready(newMsc_)) {
+        try {
+            auto vol = newMsc_.get();
+            outport_.setData(vol);
+            hasNewData_ = false;
+            getActivityIndicator().setActive(false);
+        } catch (Exception&) {
+            // Need to reset the future, VS bug:
+            // http://stackoverflow.com/questions/33899615/stdfuture-still-valid-after-calling-get-which-throws-an-exception
+            newMsc_ = {};
+            outport_.setData(nullptr);
+            hasNewData_ = false;
+            throw;
+        }
+    } else if (!newMsc_.valid()) {  // We are not waiting for a calculation
+        if (inport_.isChanged() || mscDirty_) {
+            getActivityIndicator().setActive(true);
+            updateOutport();
+        }
+    }
+}
+
+void MorseSmaleComplex::updateOutport() {
+    auto done = [this]() {
+        dispatchFront([this]() {
+            mscDirty_ = false;
+            hasNewData_ = true;
+            invalidate(InvalidationLevel::InvalidOutput);
+        });
+    };
+
+    // Save input and properties needed to calculate ttk contour tree to local variables
+    const auto inportData = inport_.getData();
+
+    // construction of ttk contour tree
+    auto compute = [this, inportData,
+                    done]() -> std::shared_ptr<const topology::MorseSmaleComplexData> {
+        ScopedClockCPU clock{"MorseSmaleComplex", "Morse-Smale complex calculation",
+                             std::chrono::milliseconds(500), LogLevel::Info};
+        auto mscData =
+            inportData->getScalarValues()
+                ->getRepresentation<BufferRAM>()
+                ->dispatch<std::shared_ptr<topology::MorseSmaleComplexData>,
+                           dispatching::filter::Scalars>([inportData](const auto buffer) {
+                    using ValueType = util::PrecisionValueType<decltype(buffer)>;
+                    using PrimitiveType = typename DataFormat<ValueType>::primitive;
+
+                    std::vector<int> offsets(inportData->getOffsets());
+
+                    ttk::MorseSmaleComplex morseSmaleComplex;
+                    morseSmaleComplex.setupTriangulation(
+                        const_cast<ttk::Triangulation*>(&inportData->getTriangulation()));
```

Comment:
Was this fixed?

Reply:
constness was only partially fixed. See e.g.
`int ttk::MorseSmaleComplex::setInputScalarField(void *const data);`